### PR TITLE
Break and re-create color tags at line break

### PR DESCRIPTION
**Pull request by dthree**
> sweet - looks good. If we just sort out this column thing I was talking about (my code does it, I'll try and get a moment to implement it in yours),

As promised :)

Looked for the most efficient way to do this while maintaining the style you did it in. This maintains linear complexity (loops through built string once more).

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -2,15 +2,17 @@
 var spliceString = require('splice-string');
 
 module.exports = function (str, cols) {
+	var pre = '';
 	var ret = '';
 	var insideEscape = false;
+	var endCode = '\u001b[39m';
```
Comment (sindresorhus) [inline index.js]: Shouldn't it support other end-codes like `49`? https://github.com/chalk/ansi-styles/blob/master/index.js

File: index.js [hunk 2]
```diff
@@ -28,11 +30,37 @@ module.exports = function (str, cols) {
...
+		} else if (insideEscape && y === 'm') {
+			insideEscape = false;
+			continue;
+		}
+
+		if (insideEscape) {
+			continue;
+		}
+
+		++visible;
```
Comment (sindresorhus) [inline index.js]: No point in using a prefix increment when you don't use the value directy: `++visible;` => `visible++;`.

File: index.js [hunk 3]
```diff
@@ -28,11 +30,37 @@ module.exports = function (str, cols) {
...
 		}
 	}
 
+	visible = 0;
+	for (var j = 0; j < pre.length; j++) {
+		var y = pre[j];
+		ret += y;
+		if (y === '\u001b') {
+			var code = parseFloat(pre.slice(j, j + 4).match(/[0-9][^m]*/));
+			escapeCode = (code === 39) ? undefined : code;
```
Comment (sindresorhus) [inline index.js]: Can be shortened to: `escapeCode = code === 39 && code;`

**Comment by sindresorhus** [comment]
// @Qix- @jbnicolai

**Comment by dthree** [comment]
K I'll throw together the fixes on these

**Comment by sindresorhus** [comment]
Can you do a quick explanation about the thinking with the implementation? Kinda hard to get that from just reading the code. And what's the reasoning for using an additional loop?

**Comment by dthree** [comment]
I might just be braindead on this subject, but I spent *way* too long trying to fit the logic into the original loop, specifically in fitting in the breaks around your splice function and getting the visible line lengths right. I can take another shot if its a big problem.

**Comment by sindresorhus** [comment]
> I can take another shot if its a big problem.

It's not, I was just curious about the reasoning ;)

**Comment by dthree** [comment]
K :) I'll fix your points and resubmit.

But yeah - nothing magical - just me. Algorithms apparently aren't my forte.

**Comment by sindresorhus** [comment]
@dthree No need to resubmit. Just push new commits to this PR ;)

File: index.js (hunk 3 above)
Comment (dthree) [inline index.js]: > Can be shortened to: `escapeCode = code === 39 && code;`

Eh? I was trying this and it wasn't producing the same result as:

`escapeCode = (code === 39) ? undefined : code; `

Am I missing something?

Comment (sindresorhus) [inline index.js]: Never mind. I was reading it incorrectly. I think it should be `null` instead of `undefined` though.

**Comment by dthree** [comment]
Yeah - files are on Ubuntu but I'm editing them on a Windows box. I guess I opened the readme and then compulsively saved nothing and it changed the permissions.

Anyway, I updated the permissions, so it should be good to go!

**Comment by silverwind** [comment]
I tink that permission thing is a Sublime Text bug related to the `atomic_save` option.

**Comment by Qix-** [comment]
I always had file permissions issues in Windows, especially when mounting and editing Ubuntu stuff - I think it's a universal thing (I've never really used Sublime).

**Comment by silverwind** [comment]
Well, I avoid these things by doing all git stuff inside Cygwin when working on Windows :)

**Comment by Qix-** [comment]
Again, though, it doesn't matter if you're mounting on a unix-y system and editing from Windows. It's the underlying system that changes the permissions to higher, I believe. Just gotta be vigilant about it :)

**Comment by dthree** [comment]
:)

Nailed it. Using Sublime.

**Comment by dthree** [comment]
@sindresorhus do you need me to do anything else on this before merging? I'm ready to use this module in `wat` but just need this line break thing working. Thanks!

File: index.js [hunk 4]
```diff
@@ -1,16 +1,21 @@
 'use strict';
 var spliceString = require('splice-string');
 
+var sets = {
+	'0': 0,	'1': 22, '2': 22, '3': 23, '4': 24, '7': 27, '8': 28, '9': 2, '30': 39, '31': 39, '32': 39, '33': 39, '34': 39, '35': 39, '36': 39, '37': 39, '90': 3, '40': 49, '41': 49, '42': 49, '43': 49, '44': 49, '45': 49, '46': 49, '47': 4
+};
```
Comment (Qix-) [inline index.js]: Why not use `ansi-styles`?

File: index.js [hunk 5]
```diff
@@ -28,11 +33,37 @@ module.exports = function (str, cols) {
...
 			lastSpace = 0;
 			visible = 0;
 		}
 	}
 
+	visible = 0;
+	for (var j = 0; j < pre.length; j++) {
+		var y = pre[j];
+		ret += y;
+		if (y === '\u001b') {
```
Comment (Qix-) [inline index.js]: Should also be checking for `\u009b`.

File: index.js (hunk 3 above)
Comment (Qix-) [inline index.js]: Explicit `undefined` is usually reserved for cases where `null` is a meaningful value. :)

File: index.js [hunk 6]
```diff
@@ -28,11 +33,37 @@ module.exports = function (str, cols) {
...
+			continue;
+		}
+
+		visible++;
+		if (escapeCode && sets[escapeCode]) {
+			if (pre[j + 1] === '\n') {
+				ret += '\u001b[' + sets[escapeCode] + 'm';
+			} else if (y === '\n') {
+				ret += '\u001b[' + escapeCode + 'm';
+			}
```
Comment (Qix-) [inline index.js]: This can be simplified.

Comment (sindresorhus) [inline index.js]: How? Notice that they're different.

Comment (Qix-) [inline index.js]: Could DRY up the string wrapping.

File: index.js (hunk 5 above)
Comment (dthree) [inline index.js]: @Qix- I'm not familiar with \u009b and didn't notice it elsewhere. What does it represent?

File: index.js (hunk 4 above)
Comment (dthree) [inline index.js]: I could, however, I would need to disassemble it anyway into the above set - in looking at it, it only seems to expose them as strings, and I don't really need that.

File: index.js (hunk 5 above)
Comment (sindresorhus) [inline index.js]: `\u001b` is used a lot, make it a constant at the top for clarity, same with `\u009b`.

**Comment by sindresorhus** [comment]
Just noticed that you've made all the other changed files executable too.

File: index.js [hunk 7]
```diff
@@ -28,11 +33,37 @@ module.exports = function (str, cols) {
...
 			visible = 0;
 		}
 	}
 
+	visible = 0;
+	for (var j = 0; j < pre.length; j++) {
+		var y = pre[j];
+		ret += y;
+		if (y === '\u001b') {
+			var code = parseFloat(pre.slice(j, j + 4).match(/[0-9][^m]*/));
```
Comment (sindresorhus) [inline index.js]: Can you use `exec` instead?

```
var code = parseFloat(/[0-9][^m]*/.exec(pre.slice(j, j + 4)));
```

File: index.js (hunk 3 above)
Comment (sindresorhus) [inline index.js]: `39` should be a constant too, for clarity.

**Comment by dthree** [comment]
Damn windows + sublime. Ok fixing.

File: index.js (hunk 7 above)
Comment (dthree) [inline index.js]: Yes

**Comment by dthree** [comment]
- Permissions fixed
- Replaced regex
- Dried up code assembly.

**Comment by dthree** [comment]
Confession: This is my first actual PR. Thanks for being patient.

File: index.js (hunk 4 above)
Comment (Qix-) [inline index.js]: Mmm they're [exposed as numbers](https://github.com/chalk/ansi-styles/blob/master/index.js#L5-L36).

**Comment by sindresorhus** [comment]
Cool. Landed a published as 0.2.0. Thanks @dthree. Congrats on your first PR 🎉

Btw, on the last 3 commits your username were suddenly `root`.

**Comment by dthree** [comment]
Thanks!!! :boom: imported it into Wat.

Haha - I noticed that after... was wrangling with `nvm` and somehow managed that. I'm learning...

**Comment by dthree** [comment]
Btw, somehow CI had an error on the `io.js` test. The error doesn't even make sense:

https://travis-ci.org/chalk/wrap-ansi

**Comment by sindresorhus** [comment]
Yeah, I'm gonna rerun it tomorrow. It happens sometimes. Not a problem with the code though.

**Comment by dthree** [comment]
K

**Pull request closed without merging**